Validate Input Attribute

nnra6864 opened the pull request:
Implements [ValidateInput] attribute with support for various condition types (fields, properties, methods), including static and nested ones.

Includes:
- Full test class to demonstrate and validate use cases.
- Compatibility with most Odin/Alchemy-like patterns.
- Notes on current limitations and planned improvements (see below).

Would appreciate feedback on:
- Expanding supported member types (e.g., int/string conditions).
- Supporting custom error messages and icons.
- Allowing non-optional parameters with default values.

<details>

<summary>Validation Tests</summary>

```cs
using ArtificeToolkit.Attributes;
using UnityEngine;

namespace Assets.Scripts
{
    public class TestClass
    {
        public bool      _alwaysFalse = false;
        public bool      _alwaysTrue  = true;
        public bool      AlwaysFalse                                     => false;
        public bool      AlwaysTrue                                      => true;
        public bool      FalseMethod()                                   => false;
        public bool      TrueMethod()                                    => true;
        public bool      IsNull(Transform tr)                            => tr;
        public bool      IsNullWithOptParam(Transform tr, int param = 0) => tr;
        public TestClass This()                                          => this;
    }

    public static class StaticClass
    {
        public static bool _alwaysFalse = false;
        public static bool _alwaysTrue  = true;
        public static bool AlwaysFalse                                     => false;
        public static bool AlwaysTrue                                      => true;
        public static bool FalseMethod()                                   => false;
        public static bool TrueMethod()                                    => true;
        public static bool IsNull(Transform tr)                            => tr;
        public static bool IsNullWithOptParam(Transform tr, int param = 0) => tr;
    }

    public class ValidationTest : MonoBehaviour
    {
        #region General

        [FoldoutGroup("General"), SerializeField, ValidateInput("FaLsE")]
        private Transform _falseString;

        [FoldoutGroup("General"), SerializeField, ValidateInput("TrUe")]
        private Transform _trueString;

        // [SerializeField, ValidateInput("0")]
        // private Transform _falseInt;
        //
        // [SerializeField, ValidateInput("1")]
        // private Transform _trueInt;

        public bool _alwaysFalse = false;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(_alwaysFalse))]
        private Transform _falseField;

        public bool _alwaysTrue = true;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(_alwaysTrue))]
        private Transform _trueField;

        public bool AlwaysFalse => false;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(AlwaysFalse))]
        private Transform _falseProperty;

        public bool AlwaysTrue => true;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(AlwaysTrue))]
        private Transform _trueProperty;

        public bool FalseMethod() => false;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(FalseMethod))]
        private Transform _falseMethod;

        public bool TrueMethod() => true;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(TrueMethod))]
        private Transform _trueMethod;

        public bool IsNull(Transform tr) => tr;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(IsNull))]
        private Transform _isNull;

        public bool IsNullWithOptParam(Transform tr, int param = 0) => tr;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(IsNullWithOptParam))]
        private Transform _isNullWithOptParam;

        public ValidationTest This() => this;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(IsNull))]
        private Transform _isNullWithMethod;

        #endregion

        #region Nested

        private TestClass _testClass = new();

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass._alwaysFalse")]
        private Transform _nestedFalseField;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass._alwaysTrue")]
        private Transform _nestedTrueField;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.AlwaysFalse")]
        private Transform _nestedFalseProperty;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.AlwaysTrue")]
        private Transform _nestedTrueProperty;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.FalseMethod")]
        private Transform _nestedFalseMethod;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.TrueMethod")]
        private Transform _nestedTrueMethod;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.IsNull")]
        private Transform _nestedIsNull;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.IsNullWithOptParam")]
        private Transform _nestedIsNullWithOptParam;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.This.IsNull")]
        private Transform _nestedIsNullWithMethod;

        #endregion

        #region Static Class

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass._alwaysFalse))]
        private Transform _staticClassFalseField;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass._alwaysTrue))]
        private Transform _staticClassTrueField;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.AlwaysFalse))]
        private Transform _staticClassFalseProperty;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.AlwaysTrue))]
        private Transform _staticClassTrueProperty;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.FalseMethod))]
        private Transform _staticClassFalseMethod;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.TrueMethod))]
        private Transform _staticClassTrueMethod;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.IsNull))]
        private Transform _staticClassIsNull;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.IsNullWithOptParam))]
        private Transform _staticClassIsNullWithOptParam;

        #endregion
    }
}
```

</details>

<details>

<summary>Issues</summary>

1. Error doesn't appear in validator when applied to a list.
2. Doesn't update whether error is displayed above the field till I reselect the object in order to redraw the inspector. This happens if I apply a validator to a field that's not actively being changed, e.g. I am changing a list count but validator is on transform, error won't show up till transform is rerendered.
3. Message is sometimes empty in the validator on the first run, gets shown on the 2nd run.
4. Message is empty in the inspector, even after reselecting the object.
5. Multiline messages are not nicely displayed in the validator.

</details>

Some things I would like to implement, but would appreciate your feedback first:
1. Implement support for overriding the message and icon in validation methods like [Odin](https://odininspector.com/attributes/validate-input-attribute): `private bool HasMeshRendererDynamicMessageAndType(GameObject gameObject, ref string errorMessage, ref InfoMessageType? messageType)`
2. Allow different value types, e.g. string/int, for members used as condition. This should be fairly easy to implement and might simplify the usage in certain cases. Let me know if this is not clear enough and I'll elaborate.
3. Allow validation methods to have non optional parameters. As of right now, either no params are allowed, or all but first have to be optional and first has to match the type of the field we are applying attribute to. This behavior is similar *I think* to what Alchemy and Odin do, however, I could expand it so that default values(e.g. 0 for int) are used for non optional params.

I should also mention that I implemented support for a ton of different use cases, not all are found in [Alchemy](https://annulusgames.github.io/Alchemy/articles/en/attributes/validate-input.html)/[Odin](https://odininspector.com/attributes/validate-input-attribute). Let me know if you feel like some are bloat/confusing and I'll remove them.

nnra6864 commented:
Merged main and added list to validation tests:

<details>

<summary>Validation Tests</summary>

```cs
using System.Collections.Generic;
using ArtificeToolkit.Attributes;
using UnityEngine;

namespace Assets.Scripts
{
    public class TestClass
    {
        public bool      _alwaysFalse = false;
        public bool      _alwaysTrue  = true;
        public bool      AlwaysFalse                                     => false;
        public bool      AlwaysTrue                                      => true;
        public bool      FalseMethod()                                   => false;
        public bool      TrueMethod()                                    => true;
        public bool      IsNull(Transform tr)                            => tr;
        public bool      IsNullWithOptParam(Transform tr, int param = 0) => tr;
        public TestClass This()                                          => this;
    }

    public static class StaticClass
    {
        public static bool _alwaysFalse = false;
        public static bool _alwaysTrue  = true;
        public static bool AlwaysFalse                                     => false;
        public static bool AlwaysTrue                                      => true;
        public static bool FalseMethod()                                   => false;
        public static bool TrueMethod()                                    => true;
        public static bool IsNull(Transform tr)                            => tr;
        public static bool IsNullWithOptParam(Transform tr, int param = 0) => tr;
    }

    public class ValidationTest : MonoBehaviour
    {
        #region General

        [FoldoutGroup("General"), SerializeField, ValidateInput("FaLsE")]
        private Transform _falseString;

        [FoldoutGroup("General"), SerializeField, ValidateInput("TrUe")]
        private Transform _trueString;

        private bool _alwaysFalse = false;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(_alwaysFalse))]
        private Transform _falseField;

        private bool _alwaysTrue = true;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(_alwaysTrue))]
        private Transform _trueField;

        public bool AlwaysFalse => false;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(AlwaysFalse))]
        private Transform _falseProperty;

        public bool AlwaysTrue => true;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(AlwaysTrue))]
        private Transform _trueProperty;

        public bool FalseMethod() => false;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(FalseMethod))]
        private Transform _falseMethod;

        public bool TrueMethod() => true;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(TrueMethod))]
        private Transform _trueMethod;

        public bool IsNull(Transform tr) => tr;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(IsNull))]
        private Transform _isNull;

        public bool IsNullWithOptParam(Transform tr, int param = 0) => tr;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(IsNullWithOptParam))]
        private Transform _isNullWithOptParam;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(IsNull))]
        private Transform _isNullWithMethod;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(ListValidation))]
        private List<Transform> _list;
        
        private bool ListValidation => _list.Count > 0;

        #endregion

        #region Nested

        private TestClass _testClass = new();

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass._alwaysFalse")]
        private Transform _nestedFalseField;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass._alwaysTrue")]
        private Transform _nestedTrueField;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.AlwaysFalse")]
        private Transform _nestedFalseProperty;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.AlwaysTrue")]
        private Transform _nestedTrueProperty;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.FalseMethod")]
        private Transform _nestedFalseMethod;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.TrueMethod")]
        private Transform _nestedTrueMethod;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.IsNull")]
        private Transform _nestedIsNull;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.IsNullWithOptParam")]
        private Transform _nestedIsNullWithOptParam;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.This.IsNull")]
        private Transform _nestedIsNullWithMethod;

        #endregion

        #region Static Class

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass._alwaysFalse))]
        private Transform _staticClassFalseField;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass._alwaysTrue))]
        private Transform _staticClassTrueField;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.AlwaysFalse))]
        private Transform _staticClassFalseProperty;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.AlwaysTrue))]
        private Transform _staticClassTrueProperty;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.FalseMethod))]
        private Transform _staticClassFalseMethod;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.TrueMethod))]
        private Transform _staticClassTrueMethod;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.IsNull))]
        private Transform _staticClassIsNull;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.IsNullWithOptParam))]
        private Transform _staticClassIsNullWithOptParam;

        #endregion
    }
}
```

</details>

nnra6864 commented:
Some new notes:
1. Since I implemented the `IArtifice_ArrayAppliedAttribute` interface, the attribute seems to get ran on list elements too. I am not sure if this was intentional, however, it could be a good thing. Maybe I could add an optional bool that would determine whether this attribute should be ran for the list or its elements?
![image](https://github.com/user-attachments/assets/aea3c7de-d4aa-48ca-84c4-304ff97638d2)

2. Right now, the attribute is quite slow when applied several times. I am not sure whether this is expected since I don't have much experience with other validation assets. I could look into optimizing it further if this is too slow.

Here are the new validation tests:

<details>

<summary>Validation Tests</summary>

```cs
using System.Collections.Generic;
using ArtificeToolkit.Attributes;
using UnityEngine;

namespace Assets.Scripts
{
    public class TestClass
    {
        public bool      _alwaysFalse = false;
        public bool      _alwaysTrue  = true;
        public bool      AlwaysFalse                                     => false;
        public bool      AlwaysTrue                                      => true;
        public bool      FalseMethod()                                   => false;
        public bool      TrueMethod()                                    => true;
        public bool      IsNull(Transform tr)                            => tr;
        public bool      IsNullWithOptParam(Transform tr, int param = 0) => tr;
        public TestClass This()                                          => this;
    }

    public static class StaticClass
    {
        public static bool _alwaysFalse = false;
        public static bool _alwaysTrue  = true;
        public static bool AlwaysFalse                                     => false;
        public static bool AlwaysTrue                                      => true;
        public static bool FalseMethod()                                   => false;
        public static bool TrueMethod()                                    => true;
        public static bool IsNull(Transform tr)                            => tr;
        public static bool IsNullWithOptParam(Transform tr, int param = 0) => tr;
    }

    public class ValidationTest : MonoBehaviour
    {
        #region General

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput("FaLsE")]
        private Transform _falseString;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput("TrUe")]
        private Transform _trueString;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput(nameof(_alwaysFalse))]
        private Transform _falseField;

        private bool _alwaysFalse = false;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput(nameof(_alwaysTrue))]
        private Transform _trueField;

        private bool _alwaysTrue = true;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput(nameof(AlwaysFalse))]
        private Transform _falseProperty;

        public bool AlwaysFalse => false;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput(nameof(AlwaysTrue))]
        private Transform _trueProperty;

        public bool AlwaysTrue => true;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput(nameof(FalseMethod))]
        private Transform _falseMethod;

        public bool FalseMethod() => false;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput(nameof(TrueMethod))]
        private Transform _trueMethod;

        public bool TrueMethod() => true;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(IsNull))]
        private Transform _isNull;

        public bool IsNull(Transform tr) => tr;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(IsNullWithOptParam))]
        private Transform _isNullWithOptParam;

        public bool IsNullWithOptParam(Transform tr, int param = 0) => tr;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput("false", LogType.Warning)]
        private Transform _warning;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(LogMessageValidation))]
        private string _logMessage;

        private bool LogMessageValidation(string logMessage, ref string lm)
        {
            lm = logMessage;
            return false;
        }
        
        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(LogTypeValidation))]
        private LogType _logType;

        private bool LogTypeValidation(LogType logType, ref LogType lt)
        {
         lt = logType;
         return false;
        }

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(ListValidation))]
        private List<Transform> _list;
        
        private bool ListValidation => _list.Count > 0;

        #endregion

        #region Nested

        private TestClass _testClass = new();

        [FoldoutGroup("Nested"), SerializeField, ReadOnly, ValidateInput("_testClass._alwaysFalse")]
        private Transform _nestedFalseField;

        [FoldoutGroup("Nested"), SerializeField, ReadOnly, ValidateInput("_testClass._alwaysTrue")]
        private Transform _nestedTrueField;

        [FoldoutGroup("Nested"), SerializeField, ReadOnly, ValidateInput("_testClass.AlwaysFalse")]
        private Transform _nestedFalseProperty;

        [FoldoutGroup("Nested"), SerializeField, ReadOnly, ValidateInput("_testClass.AlwaysTrue")]
        private Transform _nestedTrueProperty;

        [FoldoutGroup("Nested"), SerializeField, ReadOnly, ValidateInput("_testClass.FalseMethod")]
        private Transform _nestedFalseMethod;

        [FoldoutGroup("Nested"), SerializeField, ReadOnly, ValidateInput("_testClass.TrueMethod")]
        private Transform _nestedTrueMethod;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.IsNull")]
        private Transform _nestedIsNull;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.IsNullWithOptParam")]
        private Transform _nestedIsNullWithOptParam;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.This.IsNull")]
        private Transform _nestedIsNullWithMethod;

        #endregion

        #region Static Class

        [FoldoutGroup("Static Class"), SerializeField, ReadOnly,
         ValidateInput(nameof(StaticClass._alwaysFalse))]
        private Transform _staticClassFalseField;

        [FoldoutGroup("Static Class"), SerializeField, ReadOnly,
         ValidateInput(nameof(StaticClass._alwaysTrue))]
        private Transform _staticClassTrueField;

        [FoldoutGroup("Static Class"), SerializeField, ReadOnly,
         ValidateInput(nameof(StaticClass.AlwaysFalse))]
        private Transform _staticClassFalseProperty;

        [FoldoutGroup("Static Class"), SerializeField, ReadOnly,
         ValidateInput(nameof(StaticClass.AlwaysTrue))]
        private Transform _staticClassTrueProperty;

        [FoldoutGroup("Static Class"), SerializeField, ReadOnly,
         ValidateInput(nameof(StaticClass.FalseMethod))]
        private Transform _staticClassFalseMethod;

        [FoldoutGroup("Static Class"), SerializeField, ReadOnly,
         ValidateInput(nameof(StaticClass.TrueMethod))]
        private Transform _staticClassTrueMethod;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.IsNull))]
        private Transform _staticClassIsNull;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.IsNullWithOptParam))]
        private Transform _staticClassIsNullWithOptParam;

        #endregion
    }
}
```

</details>

nnra6864 commented:
I also noticed that this keeps coming back constantly, not sure if I am doing something wrong on my end or if it's a bug:
![image](https://github.com/user-attachments/assets/d77371c7-e449-4582-9565-48eea94313e6)

ZackPer commented:
Hello @nnra6864 !

First of all, thank you for submitting such an awesome validation attribute and having such a nice structure on your PR. 

Regarding the issues you mentioned:
"1. Error doesn't appear in validator when applied to a list": I pushed a fix for this. I had completely forgotten to check for the array property it self. Good catch! Also simplified the logic of type checking the attribute whether it implements IArtifice_ArrayAppliedAttribute.

"2. Doesn't update whether error is displayed...": This is a peculiar one. For this there is easy fix exactly since you need to track changes for the member which dictates the validation. I will give it some thought and come back to you.

"3. Message is sometimes empty in the validation"

"4. Message is empty in the inspector": I fixed this by exposing to the children of the CustomAttributeDrawer_Validator_BASE the InfoBox visual element. This allows for the child to dynamically update the message. Up until this point, the message was supposed to be rather static so I had not predicted for this. Also good catch, I will further improve it in the future.

"5. Multiline messages...": Intresting. I had always thought the inspector to hold really brief and simple information. 

Next batch of comments:
"1. Since I implemented the IArtifice_ArrayAppliedAttribute interface, the attribute seems to...": I do not think this was an expected behaviour and probably has been fixed with my recent commits.

"2. Performance issues": This is most likely caused by your `FindType` method with is invoked by the `ResolveNestedMember`. Specifically each time, for each part of the path, checking and invoking the `AppDomain.CurrentDomain.GetAssemblies()`. There are much better ways to do it, but even just caching the GetAssemplies result would improve the performance I would guess. But just to be consistent I suggest you check out `Artifice_SerializedPropertyExtensions` we provide. Its packed with tons of utility methods for reflection and SerializedProperty handling. For example GetTarget and GetFieldNested methods could really help out here but unfortunately the current implementation only works for FieldMembers and not properties or methods. But I think the parsing logic is there so it would not be hard to extend it further. So my most important feedback and comment is to move these methods `ResolveNestedMember` and `FindType` to the `Artifice_SerializedPropertyExtensions`. Then update the logic of the `ResolveNestedMember` to with or in a similar fashion to the `GetFieldNested `.

Your test file is really unresponsive on my PC so we 100% should improve on the last point :D 
After that, I do have some naming conventions and coding style feedback but its less important and can be done later, after the fundamental mechanism is finely tuned! 

Also, the change of the `[CustomEditor(typeof(Object), true), CanEditMultipleObjects]` is something you should not push and try to ignore. Unfortunately, the only real way of HARD disabling the ArtificeToolkit, is to comment out this line. Not a best practise but its effective. When the package is added into Unity, since its on the LibraryCache the change does not show on any Versioning control system. So this change is only shown when developing the ArtificeToolkit it self. 

Again, really good work! Let me know your thoughts and if I have missed something.

nnra6864 commented:
1. > "1. Error doesn't appear in validator when applied to a list": I pushed a fix for this.

Sounds good.

2. > "2. Doesn't update whether error is displayed...": This is a peculiar one. For this there is easy fix exactly since you need to track changes for the member which dictates the validation.

All good, this is a very edge case scenario I assume.

3. > "3. Message is sometimes empty in the validation"
4. > "4. Message is empty in the inspector": I fixed this by exposing to the children of the CustomAttributeDrawer_Validator_BASE the InfoBox visual element.

Sounds good to me.

5. > "5. Multiline messages...": Intresting. I had always thought the inspector to hold really brief and simple information.

That's a valid point. I tried making messages somewhat user friendly and they turned out longer. I can shorten them down to only useful details if you'd prefer that.

6. > "1. Since I implemented the IArtifice_ArrayAppliedAttribute interface, the attribute seems to...": I do not think this was an expected behaviour and probably has been fixed with my recent commits.

Most likely, I'll pull the changes and see if it's fixed.

7. > "2. Performance issues": This is most likely caused by your FindType method with is invoked by the ResolveNestedMember.

Thanks for the pointer, I'll look into that.

> But just to be consistent I suggest you check out Artifice_SerializedPropertyExtensions we provide.

Will do.

> So my most important feedback and comment is to move these methods ResolveNestedMember and FindType to the Artifice_SerializedPropertyExtensions.

Will do.

> Then update the logic of the ResolveNestedMember to with or in a similar fashion to the GetFieldNested.

Will look into that.

> Your test file is really unresponsive on my PC so we 100% should improve on the last point :D

Yes, it takes solid 5 seconds to load on my end so I'd say it's p bad.

> After that, I do have some naming conventions and coding style feedback but its less important and can be done later, after the fundamental mechanism is finely tuned!

Sounds good to me, just let me know what should be changed.

> Also, the change of the [CustomEditor(typeof(Object), true), CanEditMultipleObjects] is something you should not push and try to ignore.

Thanks for the explanation.

ZackPer commented:
Awesome work @nnra6864 !

Performance is now smooth, even with the heavy test script file. 

You were right about the issue with the offset. It was caused because I was first creating the ValidationLog and then calling IsValid. I did have the precondition in my mind that messages where static at the time but it was still wrong haha. So now it works like a charm.

I found one problem. The code currently does not support properties that are nested deeper than the serializedObject's target.

```
    [Serializable]
    class NestedTest
    {
        [ValidateInput("false", "Even nested results work")]
        public int x;
    }

    [SerializeField, FoldoutGroup] 
    private NestedTest test;
    
    [SerializeField, FoldoutGroup]
    private List<NestedTest> listOfTests;
```
   
So the fix this, the beginning of your CustomAttributeDrawer logic should find the property object and it's direct parent instead of the root object of the serializedObject.

Can you have a look at it?

nnra6864 commented:
Glad to hear you were able to fix the message ordering issue.

> So the fix this, the beginning of your CustomAttributeDrawer logic should find the property object and it's direct parent instead of the root object of the serializedObject.

Yes, I'll try to take care of that soon.

nnra6864 commented:
I am so sorry for that accidental force push, I reverted it.
I recently started using lazygit and it's so easy to perform an action by accident, almost lost 6h of work when submitting the initial PR :/
I addressed your concerns about nested fields, let me know if there's something else I should take care of.
Here are the latest validation tests:

<details>

<summary>Validation Tests</summary>

```cs
using System;
using System.Collections.Generic;
using ArtificeToolkit.Attributes;
using UnityEngine;

namespace Assets.Scripts
{
    public class TestClass
    {
        public bool      _alwaysFalse = false;
        public bool      _alwaysTrue  = true;
        public bool      AlwaysFalse                                     => false;
        public bool      AlwaysTrue                                      => true;
        public bool      FalseMethod()                                   => false;
        public bool      TrueMethod()                                    => true;
        public bool      IsNull(Transform tr)                            => tr;
        public bool      IsNullWithOptParam(Transform tr, int param = 0) => tr;
        public TestClass This()                                          => this;
    }

    public static class StaticClass
    {
        public static bool _alwaysFalse = false;
        public static bool _alwaysTrue  = true;
        public static bool AlwaysFalse                                     => false;
        public static bool AlwaysTrue                                      => true;
        public static bool FalseMethod()                                   => false;
        public static bool TrueMethod()                                    => true;
        public static bool IsNull(Transform tr)                            => tr;
        public static bool IsNullWithOptParam(Transform tr, int param = 0) => tr;
    }

    public class ValidationTest : MonoBehaviour
    {
        #region General

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput("FaLsE")]
        private Transform _falseString;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput("TrUe")]
        private Transform _trueString;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput(nameof(_alwaysFalse))]
        private Transform _falseField;

        private bool _alwaysFalse = false;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput(nameof(_alwaysTrue))]
        private Transform _trueField;

        private bool _alwaysTrue = true;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput(nameof(AlwaysFalse))]
        private Transform _falseProperty;

        public bool AlwaysFalse => false;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput(nameof(AlwaysTrue))]
        private Transform _trueProperty;

        public bool AlwaysTrue => true;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput(nameof(FalseMethod))]
        private Transform _falseMethod;

        public bool FalseMethod() => false;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput(nameof(TrueMethod))]
        private Transform _trueMethod;

        public bool TrueMethod() => true;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(IsNull))]
        private Transform _isNull;

        public bool IsNull(Transform tr) => tr;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(IsNullWithOptParam))]
        private Transform _isNullWithOptParam;

        public bool IsNullWithOptParam(Transform tr, int param = 0) => tr;

        [FoldoutGroup("General"), SerializeField, ReadOnly, ValidateInput("false", LogType.Warning)]
        private Transform _warning;

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(LogMessageValidation))]
        private string _logMessage;

        private bool LogMessageValidation(string logMessage, ref string lm)
        {
            lm = logMessage;
            return false;
        }
        
        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(LogTypeValidation))]
        private LogType _logType;

        private bool LogTypeValidation(LogType logType, ref LogType lt)
        {
         lt = logType;
         return false;
        }

        [FoldoutGroup("General"), SerializeField, ValidateInput(nameof(ListValidation))]
        private List<Transform> _list;
        
        private bool ListValidation => _list.Count > 0;

        #endregion

        #region Nested

        private TestClass _testClass = new();

        [FoldoutGroup("Nested"), SerializeField, ReadOnly, ValidateInput("_testClass._alwaysFalse")]
        private Transform _nestedFalseField;

        [FoldoutGroup("Nested"), SerializeField, ReadOnly, ValidateInput("_testClass._alwaysTrue")]
        private Transform _nestedTrueField;

        [FoldoutGroup("Nested"), SerializeField, ReadOnly, ValidateInput("_testClass.AlwaysFalse")]
        private Transform _nestedFalseProperty;

        [FoldoutGroup("Nested"), SerializeField, ReadOnly, ValidateInput("_testClass.AlwaysTrue")]
        private Transform _nestedTrueProperty;

        [FoldoutGroup("Nested"), SerializeField, ReadOnly, ValidateInput("_testClass.FalseMethod")]
        private Transform _nestedFalseMethod;

        [FoldoutGroup("Nested"), SerializeField, ReadOnly, ValidateInput("_testClass.TrueMethod")]
        private Transform _nestedTrueMethod;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.IsNull")]
        private Transform _nestedIsNull;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.IsNullWithOptParam")]
        private Transform _nestedIsNullWithOptParam;

        [FoldoutGroup("Nested"), SerializeField, ValidateInput("_testClass.This.IsNull")]
        private Transform _nestedIsNullWithMethod;

        #endregion

        #region Static Class

        [FoldoutGroup("Static Class"), SerializeField, ReadOnly,
         ValidateInput(nameof(StaticClass._alwaysFalse))]
        private Transform _staticClassFalseField;

        [FoldoutGroup("Static Class"), SerializeField, ReadOnly,
         ValidateInput(nameof(StaticClass._alwaysTrue))]
        private Transform _staticClassTrueField;

        [FoldoutGroup("Static Class"), SerializeField, ReadOnly,
         ValidateInput(nameof(StaticClass.AlwaysFalse))]
        private Transform _staticClassFalseProperty;

        [FoldoutGroup("Static Class"), SerializeField, ReadOnly,
         ValidateInput(nameof(StaticClass.AlwaysTrue))]
        private Transform _staticClassTrueProperty;

        [FoldoutGroup("Static Class"), SerializeField, ReadOnly,
         ValidateInput(nameof(StaticClass.FalseMethod))]
        private Transform _staticClassFalseMethod;

        [FoldoutGroup("Static Class"), SerializeField, ReadOnly,
         ValidateInput(nameof(StaticClass.TrueMethod))]
        private Transform _staticClassTrueMethod;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.IsNull))]
        private Transform _staticClassIsNull;

        [FoldoutGroup("Static Class"), SerializeField,
         ValidateInput(nameof(StaticClass.IsNullWithOptParam))]
        private Transform _staticClassIsNullWithOptParam;

        #endregion
        
        #region NestedAttribute

        [Serializable]
        private class NestedAttribute
        {
            [ValidateInput("false", "Even nested results work")]
            public int NestedField;
        }
        
        [SerializeField, FoldoutGroup("Nested Attribute")] 
        private NestedAttribute _nestedAttribute;
    
        [SerializeField, FoldoutGroup("Nested Attribute")]
        private List<NestedAttribute> _listOfNestedAttributes;
        
        #endregion
    }
}
```

</details>

![image](https://github.com/user-attachments/assets/9cbbd7ca-2749-4878-ab03-18efd912d546)

ZackPer reviewed (changes requested):
Awesome work overall @nnra6864 !

I have added some comments before merging, which have to do with naming conventions mostly. Let me know your thoughts and if you are against any of it.

Then another inconsistency is the code formatting. You align everything to match. Although this is how I personally used to code as well, I have moved away from it. If some other programmer who uses a formatter were to touch the file it would cause  a lot of changes. So since the project is already not formatted like this, I would recommend to auto format it to match the rest of the codebase.

nnra6864 commented on Editor/Artifice_CustomAttributeDrawers/CustomAttributeDrawer_Validators/Artifice_CustomAttributeDrawer_ValidateInputAttribute.cs:
```diff
@@ -0,0 +1,241 @@
+        Artifice_CustomAttributeDrawer_Validator_BASE
+    {
+        private         string _logMessage = "";
+        public override string LogMessage => _logMessage;
```
I noticed that those fields were defined with getters only in the base class so I didn't wanna mess with that.
Would you like me to also add a setter in the base class since these properties can be changed by scripts now?
As of right now, I can't turn it into auto properties afaik:
`'LogMessage.set': cannot override because 'string ArtificeToolkit.Editor.Artifice_CustomAttributeDrawers.CustomAttributeDrawer_Validators.Artifice_CustomAttributeDrawer_Validator_BASE.LogMessage' does not have an overridable set accessor`

ZackPer replied:
Ah yes you are right! It made sense then since we did not plan for dynamically changing messages. Lets leave it as is for now, and we will consider for future cases to add a simple protected set.

nnra6864 commented on Editor/Artifice_CustomAttributeDrawers/CustomAttributeDrawer_Validators/Artifice_CustomAttributeDrawer_ValidateInputAttribute.cs:
```diff
@@ -0,0 +1,241 @@
+                return false;
+            }
+
+            var validateAttribute   = fieldInfo.GetCustomAttribute<ValidateInputAttribute>();
```
Done

nnra6864 commented on Editor/Artifice_CustomAttributeDrawers/CustomAttributeDrawer_Validators/Artifice_CustomAttributeDrawer_ValidateInputAttribute.cs:
```diff
@@ -0,0 +1,241 @@
+            }
+
+            var validateAttribute   = fieldInfo.GetCustomAttribute<ValidateInputAttribute>();
+            var unresolvedCondition = validateAttribute.Condition;
```
Done

nnra6864 commented on Editor/Artifice_CustomAttributeDrawers/CustomAttributeDrawer_Validators/Artifice_CustomAttributeDrawer_ValidateInputAttribute.cs:
```diff
@@ -0,0 +1,241 @@
+        {
+            ResetValues();
+            
+            var (fieldObject, memberInfo) = Artifice_SerializedPropertyExtensions
```
Done

nnra6864 commented on Editor/Artifice_CustomAttributeDrawers/CustomAttributeDrawer_Validators/Artifice_CustomAttributeDrawer_ValidateInputAttribute.cs:
```diff
@@ -0,0 +1,241 @@
+            }
+
+            // Get nested member
+            object     validationObject;
```
Done

nnra6864 commented on Editor/Artifice_CustomAttributeDrawers/CustomAttributeDrawer_Validators/Artifice_CustomAttributeDrawer_ValidateInputAttribute.cs:
```diff
@@ -0,0 +1,241 @@
+            var  parameters    = validationMethod.GetParameters();
+            var  paramValues   = new object[parameters.Length];
+            bool assignedField = false, assignedMessage = false, assignedType = false;
+            int  i             = 0;
```
Done

nnra6864 commented on Editor/Artifice_CustomAttributeDrawers/CustomAttributeDrawer_Validators/Artifice_CustomAttributeDrawer_ValidateInputAttribute.cs:
```diff
@@ -0,0 +1,241 @@
+            var (fieldObject, memberInfo) = Artifice_SerializedPropertyExtensions
+                .ResolveNestedMember(property.propertyPath, property.serializedObject.targetObject);
+            var fieldInfo = (FieldInfo)memberInfo;
+            //object fieldObject     = property.serializedObject.targetObject;
```
Done

nnra6864 commented on Editor/Artifice_CustomAttributeDrawers/CustomAttributeDrawer_Validators/Artifice_CustomAttributeDrawer_ValidateInputAttribute.cs:
```diff
@@ -0,0 +1,241 @@
+            //object fieldObject     = property.serializedObject.targetObject;
+            var    fieldObjectType = fieldObject.GetType();
+            var    fieldName       = property.name;
+            //var fieldInfo = fieldObjectType.GetField(fieldName,
```
Done

ZackPer commented:
Hello @nnra6864 !

So.. I just update the code to be able to reevaluate (by default true) the validation in the inspector by tracking the whole serializedObject for changes. I was conflicted at first but it makes sense, since practically any change in logic could potentially alter the validation.

Also, it seems I was wrong with the `(ValidateInputAttribute)Attribute` cast. For the Validator there exists the case were the Attribute has not been set yet. Sorry for that!

I think we are almost ready. Can you test one last time with the added reevaluation feature and try to break it? If it does not break we merge it.

nnra6864 commented:
Hey, I just tested it out with all the use cases and it appears to be working flawlessly.
Great solution for updating validation when not applied to the changed field.
